content/knowledge-base/guides/write-a-composition-function-in-go.md

```diff
@@ -0,0 +1,829 @@
+function-xbuckets. When you run this command the CLI initializes your function
+using the
+[function-template-go](https://github.com/crossplane/function-template-go)
+GitHub repository as a template.
+
+```shell
+crossplane beta xpkg init function-xbuckets function-template-go -d function-xbuckets
+Initialized package "function-xbuckets" in directory "/home/negz/control/negz/function-xbuckets" from https://github.com/crossplane/function-template-go/tree/91a1a5eed21964ff98966d72cc6db6f089ad63f4 (main)
+```
+
+The `crossplane beta init xpkg` command creates a directory named
+`function-xbuckets`. When you run the command the new directory should look like
+this:
+
+```shell
+ls function-xbuckets
+Dockerfile  fn.go  fn_test.go  go.mod  go.sum  input/  LICENSE  main.go  package/  README.md  renovate.json
+```
+
+The `fn.go` file is where you add the function's code. It's useful to know about
+some other files in the template:
+
+* `main.go` runs the function. You don't need to edit `main.go`.
+* `Dockerfile` builds the function runtime. You don't need to edit `Dockerfile`.
+* The `input` directory defines the function's input type.
+* The `package` directory contains metadata used to build the function package.
+
+{{<hint "tip">}}
+<!-- vale gitlab.FutureTense = NO -->
+<!--
+This tip talks about future plans for Crossplane.
+-->
+In v1.14 of the Crossplane CLI `crossplane beta xpkg init` just clones a
+template GitHub repository. In a future release the command will automate tasks
+like replacing the template name with the new function's name. See Crossplane
+issue [#4941](https://github.com/crossplane/crossplane/issues/4941) for details.
+<!-- vale gitlab.FutureTense = YES -->
+{{</hint>}}
+
+You must make some changes before you start adding code:
+
+* Edit `package/crossplane.yaml` to change the package's name.
+* Edit `go.mod` to change the Go module's name.
+* Delete the `input` and `package/input` directories.
+
+Name your package `function-xbuckets`.
+
+The name of your module depends on where you want to keep your function code. If
+you push Go code to GitHub, you can use your GitHub username. For example
+`module github.com/negz/function-xbuckets`.
+
+The `input` directory defines a Go struct that a function can use to take input,
+using the `input` field from a Composition. The
+[composition functions]({{<ref "../../master/concepts/composition-functions">}})
+documentation explains how to pass an input to a composition function.
+function-xbuckets doesn't use an input, so you should delete the `input`
+directory.
+
+You should also delete the `package/input` directory. It contains an OpenAPI
+schema automatically generated from the structs in the `input` directory.
+
+{{<hint "tip">}}
+If you're writing a function that does use an input type, rename the type from
+`Input` to something more specific to your function. For example
+function-patch-and-transform names its input type `Resources`. Rename the API
```

Comment:
ohhhh that makes a ton of sense now to connect `resources` from P&T to the concept of `Input` 🤯